### PR TITLE
NPC activity rework - farming and generic multi-activity loop

### PR DESCRIPTION
**Pull request by ghost**
#### Summary
```SUMMARY: Features "NPC activity rework - farming and generic multi-activity loop"```

#### Purpose of change
Adds a generic multi-activity function that other activities can be plugged into without much duplication.

#### Describe the solution
In the previous incarnation of this, #31696, I basically just copied the structure of the other activity loops, and did a custom fetching loop to get tools required and so on.
Bevapdin reviewed and correctly suggested that there is too much duplication and a generic fetch_required activity could be used for a number of different activities.
So here it is.

The loop goes like this :

1. Get list of tiles to travel to, depending on activity type - farming will be farm zones, construction will be blueprint zones etc.
2. Discount any tiles that are going to be impossible, if its too dark, dangerous stuff there.
3. loop through each tile.
4. Work out if its going to be possible to do the task there.
5. if its possible but you need some items, start a fetch task , and build a requirement_id detailing what is required, for crafting or construction itll be that objects own requirements file, for other tasks ( tilling soil requires a shovel for example ) - then just form a custom requirement_id with that one item/quality/tool.
5.a if its a fetch task thats started, go back to 1, search in loot zones for the required items, build a mental map of where all these items are, and how much of each to get from each tile.
6. path to the destination.
7. once arrived there, do the activity, if its a fetch task, then move the requirement_id items, near to the original work spot, or pick it up if its a task like tilling/planting.
8. loop back if theres a backlog activity to start, or other tiles to consider.
9. once this is all done, start a tidy_up activity, where all items picked up or moved and leftover, will be placed into a nearby unsorted loot zone( harvested grown plants, a shovel they picked up etc )

The practical upshot of this, and what ive rolled into it so far, is construction , farming, fetching, and loot sorting.

Now, the NPCs do not need the components placed near the construction blueprint, they can go and fetch them from nearby loot zones, they will go and fetch seeds, shovels etc for farming tasks, and place the remainder back afterwards.

This will make it much easier going forward to enable NPC crafting, NPC butchery, log chopping , any thing you can think of, because I can just insert the applicable conditions into the loop, and the requirements for what the NPC needs, and theyll go and fetch and do it.

This will also make it easier to enable a future NPC "job " system, where they periodically scan for things that are possible to do.

#### Additional context
Current limitations - I tried to test everything I could, though its certainly possible there are edge cases where certain constructions might cause weirdness.
I think I managed to avoid all potential cases of infinite looping, though this kinda thing has a lot of potential for infinite looping to occur, it *should* fail gracefully and just stop activity, when things arent possible though.

this has a follow-on PRs that relies upon this one.

#33091 

A little gif of the NPC getting requirements for a construction from seperate loot tiles.

https://gfycat.com/whichadoredindianelephant

### CODE REVIEW

File: src/activity_handlers.h [hunk 1]
```diff
@@ -8,12 +8,15 @@
 #include <vector>
 #include <list>
 
+#include "item.h"
```
Comment (BevapDin) [inline src/activity_handlers.h]: This include appears to be unnecessary.

File: src/activity_item_handling.cpp [hunk 2]
```diff
@@ -932,433 +939,1123 @@ static construction check_build_pre( const construction &con )
     return pre_con;
 }
 
-void activity_on_turn_blueprint_move( player_activity &, player &p )
+static bool character_has_skill_for( const player *p, const construction &con )
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]:
```suggestion
static bool character_has_skill_for( const player &p, const construction &con )
```
No need for a pointer at all. Calling the function with a null pointer will just crash, so why even allow it?

File: src/activity_item_handling.cpp [hunk 3]
```diff
@@ -932,433 +939,1123 @@ static construction check_build_pre( const construction &con )
...
+            if( mgr.has( zone_type_id( "LOOT_IGNORE" ), g->m.getabs( elem ) ) ||
+                g->m.dangerous_field_at( elem ) ||
+                !g->m.can_put_items_ter_furn( elem ) ) {
+                continue;
             }
-            const std::vector<tripoint> route = route_adjacent( p, src_loc );
-            if( route.empty() ) {
-                // can't get there, can't do anything, skip it
+        }
+        for( const auto elem2 : g->m.i_at( elem ) ) {
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]:
```suggestion
        for( const auto &elem2 : g->m.i_at( elem ) ) {
```
Copying the item is just going to slow the game down.

File: src/activity_item_handling.cpp [hunk 4]
```diff
@@ -932,433 +939,1123 @@ static construction check_build_pre( const construction &con )
...
+            if( in_loot_zones && elem2.made_of_from_type( LIQUID ) ) {
                 continue;
             }
-            p.set_destination( route, player_activity( act_multiple_construction ) );
-            return;
+            if( p.backlog.front().id() == activity_id( "ACT_MULTIPLE_FARM" ) ||
+                activity_to_restore == activity_id( "ACT_MULTIPLE_FARM" ) ) {
+                // this fetch task will need to pick up an item. so check for its weight/volume before setting off.
+                if( in_loot_zones && ( elem2.type->volume > volume_allowed ||
+                                       elem2.type->weight > weight_allowed ) ) {
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]: Why are you using `type->volume` (and weight)? This is often the wrong value (e.g. its the weight/volume independent of charges, of contained items, on modifications, ...). Use `item::volume()` and `weight()`.

File: src/activity_item_handling.cpp [hunk 5]
```diff
@@ -932,433 +939,1123 @@ static construction check_build_pre( const construction &con )
...
-        // dont go there if it's dangerous.
-        bool dangerous_field = false;
-        for( const std::pair<const field_type_id, field_entry> &e : g->m.field_at( src_loc ) ) {
-            if( p.is_dangerous_field( e.second ) ) {
-                dangerous_field = true;
-                break;
+        if( !in_loot_zones ) {
+            if( const cata::optional<vpart_reference> vp = g->m.veh_at( elem ).part_with_feature( "CARGO",
+                    false ) ) {
+                vehicle *src_veh = &vp->vehicle();
```
Comment (BevapDin) [inline src/activity_item_handling.cpp]:
```suggestion
                vehicle &src_veh = vp->vehicle();
```
Again: what's with the pointers?

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by sayke** [comment]
This is awesome. Thank you for taking initiative on it!

**Comment by davidpwbrown** [comment]
There was some issue I discovered with a farming task to fetch a shovel, with the loot zone being very near the work spot , meaning that spot was discounted, like it should be for construction tasks.

So nothing was fetched.

also enacted some clang tidy suggestions, but not all, the recommendation about using a reference on the requirement id , im not totally sure about yet, will test further.

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by kevingranade** [comment]
This pull request has been mentioned on **Cataclysm: Dark Days Ahead**. There might be relevant details there:

https://discourse.cataclysmdda.org/t/npc-faction-camp-issues/21020/4

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by davidpwbrown** [comment]
jenkins rebuild

**Comment by kevingranade** [comment]
Tried to start a construction (brick wall)  and immediately crashed, I had merged about 20 other PRs at the same time, so it might be because of something else, but yea it's probably this.

**Comment by davidpwbrown** [comment]
> Tried to start a construction (brick wall) and immediately crashed, I had merged about 20 other PRs at the same time, so it might be because of something else, but yea it's probably this.

will look into it.

**Comment by davidpwbrown** [comment]
> > Tried to start a construction (brick wall) and immediately crashed, I had merged about 20 other PRs at the same time, so it might be because of something else, but yea it's probably this.
> 
> will look into it.

can't reproduce with player manually building a brick wall, or NPC doing so.

**Comment by davidpwbrown** [comment]
travis rebuild

**Pull request merged**